### PR TITLE
move padding to inference

### PR DESCRIPTION
**Pull request by acisseJZhong**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4045
* __->__ #4100

## Summary

  Move uneven sequence padding out of the common MoE model and into the vLLM inference path, where
  uneven decode batches originate. The custom vLLM V1 runner pads dense/expert-SP token counts to the
  TP degree and filters invalid expert-SP CUDA graph capture sizes.

This keeps inference-specific padding out of shared model and training code.

  ## Test Plan

  CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc-per-node=4 -m pytest -q torchtitan/
  experiments/rl/tests/test_generator.py::test_vllm_uneven_decode_tp_padding

### CODE REVIEW

File: torchtitan/models/common/moe.py
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:1]: pleasing

File: torchtitan/experiments/rl/models/vllm_worker.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class TorchTitanGPUWorker(GPUWorker):
+    """V1 worker that constructs :class:`TorchTitanGPUModelRunner`."""
+
+    def init_device(self):
+        if self.use_v2_model_runner:
+            raise ValueError(
+                "TorchTitan's vLLM integration requires the V1 model runner"
+            )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/vllm_worker.py:36]: shall we follow up with the vllm side fix so that we can upgrade to v2? @pzhan9

File: torchtitan/experiments/rl/actors/generator.py [hunk 2]
```diff
@@ -210,10 +213,14 @@ class VLLMCudagraphConfig:
...
-        self, *, max_num_seqs: int, max_num_batched_tokens: int | None = None
-    ) -> CompilationConfig | None:
-        """Build a vLLM ``CompilationConfig`` for ``mode``, or return ``None``
-        when CUDA graphs are disabled.
+        self,
+        *,
+        max_num_seqs: int,
+        max_num_batched_tokens: int | None = None,
+        expert_sequence_parallel_size: int = 1,
+        enable_sequence_parallel: bool = False,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py]: can we not have default values for these two args? You can move them upwards.

File: torchtitan/experiments/rl/models/vllm_registry.py [hunk 3]
```diff
@@ -68,15 +73,25 @@ class InferenceParallelismConfig:
     expert_parallel_degree: int = 1
     """Expert parallelism degree for MoE layers. 1 means disabled."""
 
+    enable_sequence_parallel: bool = False
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/vllm_registry.py:76]: does it run well when it's turned on?

File: torchtitan/experiments/rl/models/vllm_worker.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+from vllm.v1.worker.gpu_model_runner import GPUModelRunner
+from vllm.v1.worker.gpu_worker import Worker as GPUWorker
+
+
+class TorchTitanGPUModelRunner(GPUModelRunner):
+    """V1 runner that pads batches for dense and expert sequence parallelism."""
+
+    def _pad_for_sequence_parallelism(self, num_scheduled_tokens: int) -> int:
+        tp_size = self.vllm_config.parallel_config.tensor_parallel_size
+        enable_dense_sp = self.compilation_config.pass_config.enable_sp and tp_size > 1
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/vllm_worker.py:20]: passing `enable_sp` from compilation_config feels ... weird, but I guess that's how you work with vllm?

**Review by tianyu-l — COMMENTED** [review]
looks reasonable! some questions

File: torchtitan/experiments/rl/models/vllm_worker.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/experiments/rl/models/vllm_worker.py:20]: yes the reason is that `vllm_config.parallel_config` doesn't expose `enable_sequence_parallel`; dense SP is solely decided by the compile for example https://github.com/vllm-project/vllm/blob/adc1200597c0e8e024173bf38270a660b176642d/vllm/config/vllm.py#L1376, even if it is set to True, vllm will turn it off when 
    - tensor_parallel_size == 1; or
    - the hidden-size/device heuristic cannot produce sp_min_token_num.

To provide explicit control of SP, therefore I need to pass this into/override the compilation config and set sp_min_token_num.

Comment (tianyu-l) [reply torchtitan/experiments/rl/models/vllm_worker.py:20]: oh? I didn't see any `sp_min_token_num`

File: torchtitan/experiments/rl/models/vllm_registry.py (hunk 3 above)
Comment (acisseJZhong) [reply torchtitan/experiments/rl/models/vllm_registry.py:76]: yes it should work, vllm will be responsible to pad to tp_multiple.

File: torchtitan/experiments/rl/models/vllm_worker.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/experiments/rl/models/vllm_worker.py:20]: just updated

**Review by tianyu-l — APPROVED** [review]
